src/cachinglayer-new.ts

```diff
@@ -0,0 +1,433 @@
+    return new Promise((resolve, reject) => {
+      this._doUpdateNodes(paths, _processNodes, {
+        resolve: resolve,
+        reject: reject
+      });
+    });
+  }
+
+  private _doUpdateNodes(paths, _processNodes, promise) {
+    if(this._updateNodesRunning) {
+      this._updateNodesQueued.push({
+        paths: paths,
+        cb: _processNodes,
+        promise: promise
+      });
+      return;
+    } else {
+      this._updateNodesRunning = true;
+    }
+
+    this.getNodes(paths).then((nodes) => {
+      const existingNodes = deepClone(nodes);
+      const changeEvents = [];
+
+      nodes = _processNodes(paths, nodes);
+
+      for (let path in nodes) {
+        const node = nodes[path];
+        if(equal(node, existingNodes[path])) {
+          delete nodes[path];
+        } else if(isDocument(path)) {
+          if(
+            !equal(node.local.body, node.local.previousBody) ||
+            node.local.contentType !== node.local.previousContentType
+          ) {
+            changeEvents.push({
+              path: path,
+              origin: 'window',
+              oldValue: node.local.previousBody,
+              newValue: node.local.body === false ? undefined : node.local.body,
+              oldContentType: node.local.previousContentType,
+              newContentType: node.local.contentType
+            });
+          }
+          delete node.local.previousBody;
+          delete node.local.previousContentType;
+        }
+      }
+
+      this.setNodes(nodes).then(() => {
+        this._emitChangeEvents(changeEvents);
+        promise.resolve({statusCode: 200});
+      });
+    }).then(() => {
+      return Promise.resolve();
+    }, (err) => {
+      promise.reject(err);
+    }).then(() => {
+      this._updateNodesRunning = false;
+      const nextJob = this._updateNodesQueued.shift();
+      if(nextJob) {
+        this._doUpdateNodes(nextJob.paths, nextJob.cb, nextJob.promise);
+      }
+    });
+  }
+
+  private _emitChangeEvents(events: RSEvent[]) {
+    for (let i = 0, len = events.length; i < len; i++) {
+      this._emitChange(events[i]);
+      if(this.diffHandler) {
+        this.diffHandler(events[i].path);
+      }
+    }
+  }
+
+  private _getAllDescendentPaths(path: string) {
+    if(isFolder(path)) {
+      return this.getNodes([path]).then((nodes) => {
+        const allPaths = [path];
+        const latest = getLatest(nodes[path]);
+
+        const itemNames = Object.keys(latest.itemsMap);
+        const calls = itemNames.map((itemName) => {
+          return this._getAllDescendentPaths(path + itemName).then((paths) => {
+            for (let i = 0, len = paths.length; i < len; i++) {
+              allPaths.push(paths[i]);
+            }
+          });
+        });
+        return Promise.all(calls).then(() => {
+          return allPaths;
+        });
+      });
+    } else {
+      return Promise.resolve([path]);
+    }
+  }
+
+  private _getInternals() {
```

Comment:
```suggestion
  _getInternals() {
```
_getInternals() is actually used in unit testing to give access to private methods even though it's notation indicates it's private. Wheter we should unit test private methods or not is another topic (most agree that only public methods should be tested, and test the private methods by calling the public methods with appropriate arguments).